### PR TITLE
More visitors and tests for RipperCompat

### PR DESCRIPTION
**Pull request by noahgibbs**
Part of issue #2354

More support for operators. Add basic support for ranges, parentheses and negative integers. Tests for each of these.

Full Ripper compatibility will require a lot more, but it seems good to start simple.

### CODE REVIEW

**Comment by noahgibbs** [comment]
Ooh, interesting. JRuby seems to get a different unary minus symbol here, :- rather than :-@.

**Comment by enebo** [comment]
@noahgibbs can you tell me which case is wrong in JRuby?   I assume some incompatability in what we currently ship?

**Comment by noahgibbs** [comment]
@enebo The change I had to make was with Ripper's unary minus. Specifically, in CRuby:

```ruby
noah@Noahs-MBP-3 prism % irb
irb(main):001> require "ripper"
=> false
irb(main):002> Ripper.sexp_raw("-7")
=> [:program, [:stmts_add, [:stmts_new], [:unary, :-@, [:@int, "7", [1, 1]]]]]
```

For JRuby, it looks like the same expression is getting the symbol colon-minus rather than colon-minus-at just after :unary. Trying this locally on JRuby 9.4.3.0 (it's what I already have installed), that's indeed what I see:

```ruby
noah@Noahs-MBP-3 prism % chruby jruby-9.4.3.0
noah@Noahs-MBP-3 prism % irb
irb(main):001:0> require "ripper"
=> false
irb(main):002:0>  Ripper.sexp_raw("-7")
=> [:program, [:stmts_add, [:stmts_new], [:unary, :-, [:@int, "7", [1, 1]]]]]
```

**Comment by enebo** [comment]
Thanks @noahgibbs.  I filed an issue for this: https://github.com/jruby/jruby/issues/8091

File: lib/prism/ripper_compat.rb [hunk 1]
```diff
@@ -133,8 +167,22 @@ def visit_imaginary_node(node)
...
+      text = node.slice
+      loc = node.location
+      if text[0] == "-"
+        # TODO: This is ugly. We test that a newline between - and 7 doesn't give an error, but this could still be bad somehow.
+        bounds_line_col(loc.start_line, loc.start_column + 1)
+        on_int_val = on_int(text[1..-1])
+        bounds(node.location)
+        if RUBY_ENGINE == "jruby"
+          on_unary(:-, on_int_val)
+        else
```
Comment (eregon) [inline lib/prism/ripper_compat.rb:179]: I would suggest to run `test/prism/ripper_compat_test.rb` only on CRuby (`return unless RUBY_ENGINE == "ruby"`) and match only CRuby's behavior here.
Any difference of Ripper on JRuby/TruffleRuby is most likely an unintended incompatibility with CRuby.

**Comment by enebo** [comment]
@noahgibbs Already fixed and will be in JRuby 9.4.6.0 which will be out this month.  I appreciate you running JRuby for these tests as you will help me find some bugs :smile:

**Comment by eregon** [comment]
@enebo Ah maybe I spoke too son then :sweat_smile: 
But do you think it's good for RipperCompat to mimic JRuby's differences/incompatibilities in Ripper?
Or rather the incompatible test(s) should be skipped on JRuby then?

FWIW in TruffleRuby we plan to replace Ripper with Prism::RipperCompat once it's complete enough.
For JRuby that would avoid having to maintain Ripper too.

**Comment by noahgibbs** [comment]
What I'm doing short-term is to take Tom at his word. The next update to RipperCompat will check and only do this if JRUBY_VERSION is less than 9.4.6.0. The same approach will work if you integrate RipperCompat: it can stop bothering with the quirks for any version of JRuby or TruffleRuby that defaults to using RipperCompat :-)

If there turn out to be a lot of quirks, or quirks that are inconvenient to work around, I may revisit the approach. But differences of this size aren't a big deal. We *do* want this to be standard for Ruby tooling, which means supporting currently-in-existence versions of JRuby, TruffleRuby and pre-Prism CRuby. Presumably we figure out guidelines for how long to support these things when we're further along.

**Pull request merged**